File: code/python/generators_yield.py
"""Generators produce values on demand, so a sequence can be endless."""

from collections.abc import Iterable, Iterator
from itertools import islice


def fibonacci() -> Iterator[int]:
    previous, current = 0, 1
    while True:
        yield current
        previous, current = current, previous + current


def traced(source: Iterable[int], log: list[int]) -> Iterator[int]:
    for value in source:
        log.append(value)
        yield value


def read_paragraphs(text: str) -> Iterator[str]:
    """Groups lines into paragraphs without holding the whole text twice."""
    block: list[str] = []
    for line in text.splitlines():
        if line.strip():
            block.append(line.strip())
        elif block:
            yield " ".join(block)
            block = []
    if block:
        yield " ".join(block)


print(list(islice(fibonacci(), 12)))

log: list[int] = []
first_three = list(islice((n for n in traced(range(1, 1000), log) if n % 7 == 0), 3))
print("multiples of seven:", first_three)
print(f"the source was pulled {len(log)} times, not 999")

TEXT = """First line
second line

A new paragraph
wrapped over two lines
"""
for paragraph in read_paragraphs(TEXT):
    print("-", paragraph)


# yield from delegates to another iterable.
def all_lines() -> Iterator[str]:
    yield "Amber"
    yield from ["Cobalt", "Emerald"]


print(list(all_lines()))


# send() passes a value back in at the yield.
def accumulator() -> Iterator[int]:
    total = 0
    while True:
        added = yield total
        total += added or 0


running = accumulator()
next(running)  # advance to the first yield
print(running.send(10), running.send(5), running.send(100))


# A generator's cleanup runs when it is closed or garbage collected.
def with_cleanup() -> Iterator[int]:
    try:
        yield 1
        yield 2
    finally:
        print("generator cleaned up")


values = with_cleanup()
print(next(values))
values.close()

# A generator expression is the same thing without the def.
squares = (n * n for n in range(10))
print("sum of squares:", sum(squares))
print("exhausted, so nothing is left:", list(squares))

File: code/php/generators-yield.php
<?php

declare(strict_types=1);

// A generator produces values one at a time, so a sequence can be endless
// or larger than memory.

function fibonacci(): Generator
{
    [$previous, $current] = [0, 1];

    while (true) {
        yield $current;
        [$previous, $current] = [$current, $previous + $current];
    }
}

function take(iterable $source, int $count): Generator
{
    $taken = 0;
    foreach ($source as $value) {
        if ($taken++ >= $count) {
            return;
        }
        yield $value;
    }
}

/** Reads a file one line at a time instead of loading all of it. */
function readLines(string $path): Generator
{
    $handle = fopen($path, 'rb');
    if ($handle === false) {
        throw new RuntimeException("cannot open {$path}");
    }
    try {
        while (($line = fgets($handle)) !== false) {
            yield rtrim($line, "\r\n");
        }
    } finally {
        fclose($handle);
    }
}

echo implode(' ', iterator_to_array(take(fibonacci(), 12))), PHP_EOL;

// Keys can be yielded too.
function zones(): Generator
{
    yield 'Alder Cross' => 2;
    yield 'Quill Wharf' => 3;
    yield 'Saltwick Halt' => 5;
}

foreach (zones() as $station => $zone) {
    echo "  {$station} sits in zone {$zone}", PHP_EOL;
}

// yield from delegates to another iterable, keeping the laziness.
function allLines(): Generator
{
    yield 'Amber';
    yield from ['Cobalt', 'Emerald'];
    return 'done'; // available through getReturn()
}

$lines = allLines();
foreach ($lines as $line) {
    echo "  {$line}", PHP_EOL;
}
echo 'return value: ', $lines->getReturn(), PHP_EOL;

// send() passes a value back into the generator at the yield.
function accumulator(): Generator
{
    $total = 0;
    while (true) {
        $added = yield $total;
        $total += $added ?? 0;
    }
}
$sum = accumulator();
$sum->current();
echo $sum->send(10), ' ', $sum->send(5), ' ', $sum->send(100), PHP_EOL;

// Reading a file lazily: memory stays flat however large the file is.
$path = tempnam(sys_get_temp_dir(), 'sample-assets-');
file_put_contents($path, implode(PHP_EOL, [
    'station,line,zone',
    'Alder Cross,Amber,2',
    'Quill Wharf,Cobalt,3',
    'Saltwick Halt,Amber,5',
]));

$total = 0;
foreach (take(readLines($path), 100) as $index => $line) {
    if ($index === 0) {
        continue;
    }
    $total += (int) explode(',', $line)[2];
}
echo "zones add up to {$total}", PHP_EOL;
unlink($path);
